packages/libp2p/src/connection-manager/utils.ts

```diff
@@ -28,3 +29,35 @@ export async function resolveMultiaddrs (ma: Multiaddr, options: ResolveOptions
 
   return output
 }
+
+/**
+ * Converts a multiaddr string or object to an IpNet object.
+ * If the multiaddr doesn't include /ipcidr, it will encapsulate with the appropriate CIDR:
+ * - /ipcidr/32 for IPv4
+ * - /ipcidr/128 for IPv6
+ *
+ * @param {string | Multiaddr} ma - The multiaddr string or object to convert.
+ * @returns {IpNet} The converted IpNet object.
+ * @throws {Error} Throws an error if the multiaddr is not valid.
+ */
+export function multiaddrToIpNet (ma: string | Multiaddr): IpNet {
+  try {
+    let parsedMa: Multiaddr
+    if (typeof ma === 'string') {
+      parsedMa = multiaddr(ma)
+    } else {
+      parsedMa = ma
+    }
+
+    // Check if /ipcidr is already present
+    if (!parsedMa.protoNames().includes('ipcidr')) {
+      const isIPv6 = parsedMa.protoNames().includes('ip6')
```

Comment:
Can avoid creating the proto names array twice, also just check the first proto for `ip6` to handle future craziness like `/ip4/.../tcp/1234/tunnel/ip6/.../udp/123/p2p/Qmfoo`:

```suggestion
    const protoNames = parsedMa.protoNames()

    // Check if /ipcidr is already present
    if (!protoNames.includes('ipcidr')) {
      const isIPv6 = protoNames[0] === 'ip6'
```

Reply:
Yep! 😄